pyrs/transpiler.py

```diff
@@ -545,6 +546,49 @@ def visit_Assign(self, node):
                 )
 
             return f"{kw} {self.visit(target)} = vec![{elements}];"
+        elif isinstance(node.value, ast.Set):
+            types = get_element_types(node.value.elts, node.scopes, self._type_map)
+            if len(set(types)) == 1 and types[0] is not None:
+                typename = types[0]
+            else:
+                typename = "_"
+
+            target = self.visit(target)
+            value = self.visit(node.value)
+
+            if kw in ["const", "static"]:
+                self._usings.add("lazy_static::lazy_static")
+                if typename == "&str":
+                    typename = "&'static str"
+                return f"lazy_static! {{ static ref {target}: HashSet<{typename}> = {value}; }}"
+
+            return f"{kw} {target} = {value};"
+        elif isinstance(node.value, ast.Dict):
+            key_types = get_element_types(node.value.keys, node.scopes, self._type_map)
+            if len(set(key_types)) == 1 and key_types[0] is not None:
+                key_typename = key_types[0]
+            else:
+                key_typename = "_"
+            value_types = get_element_types(
+                node.value.values, node.scopes, self._type_map
+            )
+            if len(set(value_types)) == 1 and value_types[0] is not None:
+                value_typename = value_types[0]
+            else:
+                value_typename = "_"
+
```

Comment:
This logic is already incorporated into type inference. What does `node.annotation` return?

Reply:
Now I've got the test case rebuilt, I'll check this out.

Reply:
`node.annotation` and `node.value.annotation` in the context of visit_Assign didn't have anything for these.
However the implementation of `get_element_types` has now become much simpler.

Reply:
The previous version of `get_element_types` could infer that:

```
def foo(a: int):
    c = {}
    d = 1
    if a < 10:
        b = d
        if b > 0:
            c = {b: 3}
 
    return c    
```

c is a `Dict[int, int]`. The variant in this diff can only do that for a limited number of cases (in that it doesn't chase down the definition of `b`).

But I do agree

Reply:
I do not mind that this `get_element_types` isnt as good as the previous one, especially if you are going to re-write it afterwards.  `get_element_types` is only used to establish whether all members of a global dict/set/list have the same type.  It isnt supposed to be used in any other way.  Your code sample above is not what this function is being used for.

IMO the code only needs to pass the tests that we've established.  Otherwise progress is going to be much slower and more frustrating.

And thus IMO it is counter-productive to merge only the first two commits here - the third commit has the tests for the first two.  The third commit provides the basis on which you can further enhance the type inference, and ideally you add more tests.